dll/win32/shell32/dialogs/filedefext.cpp

```diff
@@ -885,11 +885,256 @@ CFileDefExt::VersionPageProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lPar
+void CFileDefExt::OnFolderCustChangeIcon(HWND hwndDlg)
+{
+    WCHAR szPath[MAX_PATH];
+    INT nIconIndex;
+
+    // m_szFolderIconPath, m_nFolderIconIndex --> szPath, nIconIndex
+    if (m_szFolderIconPath[0])
+    {
+        StringCchCopyW(szPath, _countof(szPath), m_szFolderIconPath);
+        nIconIndex = m_nFolderIconIndex;
+    }
+    else
+    {
+        szPath[0] = 0;
+        nIconIndex = 0;
+    }
+
+    // let the user choose the icon
+    if (PickIconDlg(hwndDlg, szPath, _countof(szPath), &nIconIndex))
+    {
+        // changed
+        m_bFolderIconIsSet = TRUE;
+        PropSheet_Changed(GetParent(hwndDlg), hwndDlg);
+
+        // update
+        StringCchCopyW(m_szFolderIconPath, _countof(m_szFolderIconPath), szPath);
+        m_nFolderIconIndex = nIconIndex;
+        UpdateFolderIcon(hwndDlg);
+    }
+}
+
+// IDD_FOLDER_CUSTOMIZE PSN_APPLY
+BOOL CFileDefExt::OnFolderCustApply(HWND hwndDlg)
+{
+    // build the desktop.ini file path
+    WCHAR szIniFile[MAX_PATH];
+    StringCchCopyW(szIniFile, _countof(szIniFile), m_wszPath);
+    PathAppendW(szIniFile, L"desktop.ini");
+
+    if (m_bFolderIconIsSet)     // it is set!
+    {
+        DWORD attrs;
+
+        // change folder attributes
+        attrs = GetFileAttributesW(m_wszPath);
+        attrs &= ~(FILE_ATTRIBUTE_SYSTEM | FILE_ATTRIBUTE_READONLY);
+        SetFileAttributesW(m_wszPath, attrs);
+
+        // change desktop.ini attributes
+        attrs = GetFileAttributesW(szIniFile);
+        attrs &= ~(FILE_ATTRIBUTE_SYSTEM | FILE_ATTRIBUTE_HIDDEN | FILE_ATTRIBUTE_READONLY);
+        SetFileAttributesW(szIniFile, attrs);
+
+        if (m_szFolderIconPath[0])
+        {
+            // write IconFile and IconIndex
+            WritePrivateProfileStringW(s_szShellClassInfo, s_szIconFile, m_szFolderIconPath, szIniFile);
+
+            WCHAR szInt[32];
+            StringCchPrintfW(szInt, _countof(szInt), L"%d", m_nFolderIconIndex);
+            WritePrivateProfileStringW(s_szShellClassInfo, s_szIconIndex, szInt, szIniFile);
+
+            // flush!
+            WritePrivateProfileStringW(NULL, NULL, NULL, szIniFile);
+        }
+        else
+        {
+            // erase three values
+            WritePrivateProfileStringW(s_szShellClassInfo, s_szIconFile, NULL, szIniFile);
+            WritePrivateProfileStringW(s_szShellClassInfo, s_szIconIndex, NULL, szIniFile);
+            WritePrivateProfileStringW(s_szShellClassInfo, s_szIconResource, NULL, szIniFile);
+
+            // flush!
+            WritePrivateProfileStringW(NULL, NULL, NULL, szIniFile);
+        }
+
+        // change desktop.ini attributes
+        attrs = GetFileAttributesW(szIniFile);
+        attrs |= FILE_ATTRIBUTE_SYSTEM | FILE_ATTRIBUTE_HIDDEN;
+        SetFileAttributesW(szIniFile, attrs);
+
+        // change folder attributes
+        attrs = GetFileAttributesW(m_wszPath);
+        attrs |= FILE_ATTRIBUTE_SYSTEM;
+        SetFileAttributesW(m_wszPath, attrs);
```

Comment:
Why are you setting the folder to +system here? Is this what Windows does? It just seems like an odd change to make.

Reply:
2K3:
![attrib](https://user-images.githubusercontent.com/2107452/42048723-10b12b6e-7b3f-11e8-8fd1-5109052fd8f1.png)
The folder is +readonly. The desktop.ini is +system +hidden.

Reply:
OK.

Reply:
Does this also apply for the other folders like, "My Documents" or "My Images" ?

Reply:
Some special folders have the system attribute. The system attribute or read-only attribute are required for custom folder icon.